_drafts/reading-legacy-code-to-harmonize-ruby-and-rails-code-loaders.md

```diff
@@ -0,0 +1,701 @@
+
+<iframe width="560" height="315" src="//www.youtube.com/embed/blKRYmbpDas" frameborder="0" allowfullscreen>
+<!-- source for youtube video is in /videos/dbeugging-in-namespace-module-content.mov -->
+</iframe>
+
+# Review
+
+So, with this final version of the Metasploit Module loader, the Metasploit Pro's `prosvc` was able to use `ActiveSupport::Dependencies.autoload_paths` to load the same code as Metasploit Pro UI.  Loaded Metasploit Modules have assigned namespace names and the entire loading process is debuggable.  Let's review what we've covered.
+
+* There are 3 ways to load entire files in the Ruby standard library:
+  1. [`#load`](#load)
+  2. [`#require`](#require)
+  3. [`#autoload`](#autoload)
+* There are 2 ways to load entire files in ActiveSupport:
+  1. [`ActiveSupport::Autoload`](#activesupportautoload)
+  2. [`ActiveSupport::Dependencies.autoload_paths`](#activesupportdependenciesautoloadpaths)
+* `Module.new` creates anonymous modules
+  * They get a `Module#name` when a constant is set to the `Module`
+  * They break `ActiveSupport::Dependencies`'s `const_missing`
+* When files are too big they can be broken up
+  * Ensure there is one `Class` or `Module` per file
+  * Group related methods into `include`d `Module`s
+  * Break up hierarchies of methods into `Class` hierarchies
+* Parts of files or entire files can be loaded by `module_eval`
+  * `module_eval` captures the lexical scope only with String
+* Lexical scope influences constant look-up
+  * Nesting `module` declaration has a different lexical scope than `::` separated names
+  * The lexical scope can be retrieved with `Module.new`
+* Passing the path and line to `module_eval` allows debugging in string code.
+
+# Acknowledgements
+
+I'd like to thank [James "Egypt" Lee](https://github.com/jlee-r7) for [imparting his knowledge of how to use the various msfconsole commands to ensure my changes didn't break anything](https://github.com/rapid7/metasploit-framework/pull/843), fixing my [lambda vs proc bug](https://github.com/rapid7/metasploit-framework/commit/3a572625f5d264fff38bb83c97a76167f45ccf04), and [Ruby 1.8-incompatibility](https://github.com/rapid7/metasploit-framework/commit/10dafcd09f2764b8b3e1239752a5b8654b795af4).  I'd like to thank [Samuel "Shuckins" Huckins](https://github.com/shuckins-r7) for [testing these changes against Metasploit Pro](https://github.com/rapid7/metasploit-framework/pull/843#issuecomment-9045438).  I'd like to thank [HD Moore](https://github.com/hmoore-r7) for [spotting](https://github.com/rapid7/metasploit-framework/commit/2dce6e63472a7594bf53a6793163cd7cb088d484) when I missed automatic namespace names colliding with real Modules. I'd like to thank [Trevor Rosen](https:/github.com/trosen-r7) for allowing me to spend weeks to fix this the right way.  I'd like to thank [Regina Imhoff](https://github.com/StabbyMcDuck) for reviewing this article and accompanying presentation.
```

Comment:
I'd like to thank [James **~~"Egypt"~~** **"egypt"** Lee](https://github.com/jlee-r7) for [imparting his knowledge of how to use the various msfconsole commands to ensure my changes didn't break anything](https://github.com/rapid7/metasploit-framework/pull/843), fixing my [lambda vs proc bug](https://github.com/rapid7/metasploit-framework/commit/3a572625f5d264fff38bb83c97a76167f45ccf04), and [Ruby 1.8-incompatibility](https://github.com/rapid7/metasploit-framework/commit/10dafcd09f2764b8b3e1239752a5b8654b795af4).  I'd like to thank [Samuel "Shuckins" Huckins](https://github.com/shuckins-r7) for [testing these changes against Metasploit Pro](https://github.com/rapid7/metasploit-framework/pull/843#issuecomment-9045438).  I'd like to thank [HD Moore](https://github.com/hmoore-r7) for [spotting](https://github.com/rapid7/metasploit-framework/commit/2dce6e63472a7594bf53a6793163cd7cb088d484) when I missed automatic namespace names colliding with real Modules. I'd like to thank [Trevor Rosen](https:/github.com/trosen-r7) for allowing me to spend weeks to fix this the right way.  I'd like to thank [Regina Imhoff](https://github.com/StabbyMcDuck) for reviewing this article and accompanying presentation.